core/src/nucleus/ribosome/api/capabilities.rs

```diff
@@ -0,0 +1,83 @@
+use crate::{
+    agent::actions::commit::commit_entry,
+    nucleus::ribosome::{api::ZomeApiResult, Runtime},
+};
+use holochain_core_types::{
+    cas::content::Address,
+    entry::{cap_entries::CapTokenGrant, Entry},
+    error::HolochainError,
+};
+use holochain_wasm_utils::api_serialization::capabilities::GrantCapabilityArgs;
+use std::convert::TryFrom;
+use wasmi::{RuntimeArgs, RuntimeValue};
+
+pub fn invoke_grant_capability(runtime: &mut Runtime, args: &RuntimeArgs) -> ZomeApiResult {
+    let context = runtime.context()?;
+    // deserialize args
+    let args_str = runtime.load_json_string_from_args(&args);
+    let args = match GrantCapabilityArgs::try_from(args_str.clone()) {
+        Ok(input) => input,
+        Err(..) => return ribosome_error_code!(ArgumentDeserializationFailed),
+    };
+
+    let task_result: Result<Address, HolochainError> =
+        match CapTokenGrant::create(&args.id, args.cap_type, args.assignees, args.functions) {
+            Ok(grant) => context.block_on(commit_entry(
+                Entry::CapTokenGrant(grant.clone()),
+                None,
+                &context.clone(),
+            )),
+            Err(err) => Err(HolochainError::ErrorGeneric(format!(
+                "Unable to commit capability grant: {}",
+                err
+            ))),
+        };
+
+    runtime.store_result(task_result)
```

Comment:
This store result invocation is another cross cutting concern these api functions always seem to exhibit. We would benefit from isolating common patterns such as these and leverage a code reuse feature from rust to DRY this out (maybe a trait implementation per api function, or a macro that all functions get wrapped over).

Reply:
Yep, there's a set of SOA nodes about refactoring the way zome api calls are built...